File: 153-find-minimum-in-rotated-sorted-array/153-find-minimum-in-rotated-sorted-array.py
class Solution:
    def findMin(self, nums: List[int]) -> int:
        # find first element smaller than first
        n = len(nums)
        left, right = 0, n
        first = nums[0]
        # if first <= nums[-1]:
        #     return nums[0]
        
        while left < right:
            mid = left + (right - left)//2
            if nums[mid] >= first:
                left = mid + 1
            else:
                right = mid
                
        return nums[left % n]
            

File: 153-find-minimum-in-rotated-sorted-array/153-find-minimum-in-rotated-sorted-array.cpp
class Solution {
public:
    
    // template<n>
    // static bool smallerLast(int cur){
    //     return cur <= n;
    // }
    bool biggerLast(int cur, int last){
            return last < cur;
    }
    
    int findMin(vector<int>& nums) {
        
        int left = 0;
        int right = nums.size();
        int last = nums.back();
        
        if (nums.front() <= last) return nums[0];
            
        while (left < right){
            int mid = (left + right) / 2;
            
            if (biggerLast(nums[mid], last))
                left = mid + 1;
            
            else right = mid;
        }
        
        return nums[left];
    }
};